File: code/java/剑指offer66.java
class Solution {
    public int[] constructArr(int[] a) {
        if(a.length==0) return new int[]{};
        int[] pre = new int[a.length-1];
        int[] order = new int[a.length-1];
        pre[0] = a[0];
        for(int i=1;i<a.length-1;i++) pre[i] = pre[i-1]*a[i];
        order[a.length-2] = a[a.length-1];
        for(int i=a.length-3;i>=0;i--) order[i] = order[i+1]*a[i+1];
        int[] res = new int[a.length];

        // for(int i=0;i<pre.length;i++) System.out.println(pre[i]);
        // System.out.println();
        // for(int i=0;i<pre.length;i++) System.out.println(order[i]);

        for(int i=0;i<a.length;i++){
            if(i==0) res[i] = order[0];
            else if(i==a.length-1) res[i] = pre[a.length-2];
            else res[i] = pre[i-1]*order[i];
        }
        return res;
    }
}

File: code/python/剑指offer66.py
class Solution:
    def constructArr(self, a: List[int]) -> List[int]:
        if not a:return []
        pre = [a[0]]
        for num in a[1:-1]:
            pre.append(pre[-1]*num)
        order = [a[-1]]
        for num in a[::-1][1:-1]:
            order.append(order[-1]*num)
        order = order[::-1]
        print(pre,order)
        res = []
        for i in range(len(a)):
            if i==0:
                res.append(order[0])
            elif i==len(a)-1:
                res.append(pre[-1])
            else:
                res.append(pre[i-1]*order[i])
        return res